### PR TITLE
Adds an example approver-policy plugin

### PR DESCRIPTION
**Pull request by irbekrm**
This PR adds an example approver policy plugin.

All the code is in main.go and the Readme explains its purpose.
The main goal is to provide a working sample implementation, not a best practices usable plugin.

This is an example implementation only, for folks to understand what needs to be implemented, but it does also work, to try it out:
- create a kind cluster
- deploy cert-manager with default approver disabled
- set IMAGE_REPO var in Makefile to something
- run `make deploy`
- apply policy resources and issuer and cert from ./deploy/example

/cc @smuda

When this PR merges, I will add a link to this repo in our docs https://cert-manager.io/docs/projects/approver-policy/#plugins

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# example-approver-policy-plugin
+
+This repo contains an example [cert-manager/approver-policy plugin](https://cert-manager.io/docs/projects/approver-policy/#plugins).
+
+
+> :warning:  This is plugin is not meant to be actually used and does not contain a best practices production ready code.
```
Comment (jsoref) [inline README.md:6]: `to actually be used`

`and its code is neither production ready nor does it follow best practices`

File: README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+## Implementing a custom approver plugin
+
+[cert-manager/approver-policy](https://cert-manager.io/docs/projects/approver-policy/) can be extended via a plugin mechanism where a custom plugin can be written with custom logic for evaluating `CertificateRequest`s and `CertificateRequestPolicy`s. This can then be registered with the core cert-manager/approver-policy (in Go code) and a single image can be built that will have both the core approver and the custom plugin.
+
+To get started with writing an approver plugin you can clone this repo and change the logic in the  methods of the example plugin to do what you want it to do, add tests and the desired packaging mechanism.
```
Comment (jsoref) [inline README.md:12]: There's a double space in `the  methods`

File: README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+- build a single Go binary that contains the custom plugin(s) that you wish to use as well as the upstream approver-policy. The entrypoint should be [root command of approver-policy](https://github.com/cert-manager/approver-policy/blob/v0.6.3/cmd/main.go#L24)
+
+- package the whole project using your favourite packaging mechanism. This repo contains an example Dockerfile and an example Helm chart that includes the core cert-manager/approver-policy's Helm chart.
+
+## example plugin
+
+This repo contains an example plugin `example-approver-policy-plugin` that allows to specify a weekday when `CertificateRequest`s can be approved.
+
+See an example `CertificateRequestPolicy` that allows issuance only on Tuesdays in ./deploy/examples/tuesday.yaml:
```
Comment (jsoref) [inline README.md:28]: Can you make this a link to the file?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+spec:
+  allowed:
+    ... # Be aware that using a plugin does not disable the core approver- a CertificateRequest still has to match the allowed block here even if a plugin is specified
+  selector:
+   ...
+  plugins:
+    example-approver-policy-plugin:
+      values:
+        day: "2" # Tuesday (0 - Sunday, 1 - Monday etc)
+```
```
Comment (jsoref) [inline README.md:44]: Please include a newline at the end of the file

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+help: ## Display this help.
+	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[36m<target>\033[0m\n"} /^[a-zA-Z_0-9-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
+
+.PHONY: image
+image:
+	docker buildx build \
+	  --platform linux/amd64 \
+	  --load \
+	  -t $(IMAGE_REPO)/cert-manager-example-approver-policy-plugin:$(VERSION) \
+	  .
```
Comment (wallrj) [inline Makefile:17]: I suggest using plain `docker build`. 
I have never understood how to install buildx and I suppose there will be others like me.

**Review by wallrj — CHANGES_REQUESTED** [review]
Thanks @irbekrm 

It looks good to me.
I tried it out on my laptop and after replacing the `buildx` command, I was able to deploy it.
I agree with @jsoref 's suggestions and typo fixes, so please apply those, consider not using `buildx` and then merge.



<details>
<summary>$ kubectl logs -n cert-manager deploy/cert-manager-approver-policy</summary>
```
I0428 09:57:00.654605       1 merged_client_builder.go:121] Using in-cluster configuration
I0428 09:57:00.753561       1 webhook.go:67] webhook "msg"="running tls bootstrap process..."
I0428 09:57:00.754949       1 reflector.go:221] Starting reflector *v1.Secret (1m0s) from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:00.754998       1 reflector.go:257] Listing and watching *v1.Secret from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:00.758420       1 authority.go:324] webhook/tls/certificate-authority "msg"="Generating new root CA"
I0428 09:57:00.773623       1 authority.go:388] webhook/tls/certificate-authority "msg"="Generated new root CA"
I0428 09:57:00.774218       1 authority.go:259] webhook/tls/certificate-authority "msg"="Detected change in CA secret data, notifying watchers..."
I0428 09:57:00.855140       1 shared_informer.go:303] caches populated
I0428 09:57:01.754614       1 tls.go:235] webhook/tls "msg"="generating new ECDSA private key"
I0428 09:57:01.755021       1 tls.go:252] webhook/tls "msg"="signing new serving certificate"
I0428 09:57:01.757653       1 tls.go:258] webhook/tls "msg"="signed new serving certificate"
I0428 09:57:01.757897       1 tls.go:294] webhook/tls "msg"="updated serving TLS certificate"
I0428 09:57:01.757958       1 webhook.go:78] webhook "msg"="tls bootstrap process complete"
I0428 09:57:01.757985       1 webhook.go:91] webhook "msg"="registering webhook endpoints"
I0428 09:57:01.758154       1 cmd.go:83] main "msg"="preparing approvers..."
I0428 09:57:01.758195       1 cmd.go:85] main "msg"="preparing approver..." "approver"="allowed"
I0428 09:57:01.758206       1 cmd.go:85] main "msg"="preparing approver..." "approver"="constraints"
I0428 09:57:01.758214       1 cmd.go:85] main "msg"="preparing approver..." "approver"="example-approver-policy-plugin"
I0428 09:57:01.758224       1 cmd.go:90] main "msg"="all approvers ready..."
I0428 09:57:01.758370       1 cmd.go:101] main "msg"="starting approver-policy..."
I0428 09:57:01.758623       1 internal.go:369] controller-manager "msg"="Starting server" "addr"={"IP":"::","Port":6060,"Zone":""} "kind"="health probe"
I0428 09:57:01.758700       1 internal.go:369] controller-manager "msg"="Starting server" "addr"={"IP":"::","Port":9402,"Zone":""} "kind"="metrics" "path"="/metrics"
I0428 09:57:01.758792       1 shared_informer.go:303] caches populated
I0428 09:57:01.758883       1 tls.go:152] webhook/tls "msg"="starting webhook tls manager"
I0428 09:57:01.759044       1 leaderelection.go:248] attempting to acquire leader lease cert-manager/policy.cert-manager.io...
I0428 09:57:01.768312       1 leaderelection.go:258] successfully acquired lease cert-manager/policy.cert-manager.io
I0428 09:57:01.768529       1 recorder.go:103] controller-manager/events "msg"="cert-manager-approver-policy-84fc6c7999-k8fs4_c0dada0a-2c74-4a2a-a5ea-418df3f7d766 became leader" "object"={"kind":"Lease","namespace":"cert-manager","name":"policy.cert-manager.io","uid":"638e0327-bf93-42af-9087-9ab0722f5125","apiVersion":"coordination.k8s.io/v1","resourceVersion":"2168"} "reason"="LeaderElection" "type"="Normal"
I0428 09:57:01.768537       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest" "source"="kind source: *v1.CertificateRequest"
I0428 09:57:01.768599       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest" "source"="kind source: *v1alpha1.CertificateRequestPolicy"
I0428 09:57:01.768638       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest" "source"="kind source: *v1.PartialObjectMetadata"
I0428 09:57:01.768661       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest" "source"="kind source: *v1.PartialObjectMetadata"
I0428 09:57:01.768674       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequestpolicy" "controllerGroup"="policy.cert-manager.io" "controllerKind"="CertificateRequestPolicy" "source"="kind source: *v1alpha1.CertificateRequestPolicy"
I0428 09:57:01.768695       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest" "source"="kind source: *v1.PartialObjectMetadata"
I0428 09:57:01.768712       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest" "source"="kind source: *v1.PartialObjectMetadata"
I0428 09:57:01.768726       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequestpolicy" "controllerGroup"="policy.cert-manager.io" "controllerKind"="CertificateRequestPolicy" "source"="channel source: 0xc000598cd0"
I0428 09:57:01.768732       1 controller.go:186] controller-manager "msg"="Starting EventSource" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest" "source"="kind source: *v1.PartialObjectMetadata"
I0428 09:57:01.768752       1 controller.go:194] controller-manager "msg"="Starting Controller" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest"
I0428 09:57:01.768775       1 controller.go:194] controller-manager "msg"="Starting Controller" "controller"="certificaterequestpolicy" "controllerGroup"="policy.cert-manager.io" "controllerKind"="CertificateRequestPolicy"
I0428 09:57:01.768938       1 reflector.go:221] Starting reflector *v1alpha1.CertificateRequestPolicy (9h51m22.962312966s) from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.768963       1 reflector.go:257] Listing and watching *v1alpha1.CertificateRequestPolicy from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.768938       1 reflector.go:221] Starting reflector *v1.PartialObjectMetadata (9h3m32.031556927s) from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769064       1 reflector.go:221] Starting reflector *v1.PartialObjectMetadata (9h53m4.822353028s) from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769094       1 reflector.go:257] Listing and watching *v1.PartialObjectMetadata from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769102       1 reflector.go:221] Starting reflector *v1.CertificateRequest (9h36m29.427011127s) from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769118       1 reflector.go:257] Listing and watching *v1.CertificateRequest from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769067       1 reflector.go:257] Listing and watching *v1.PartialObjectMetadata from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769300       1 reflector.go:221] Starting reflector *v1.PartialObjectMetadata (10h18m25.734628022s) from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769333       1 reflector.go:257] Listing and watching *v1.PartialObjectMetadata from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769581       1 reflector.go:221] Starting reflector *v1.PartialObjectMetadata (9h36m3.077577405s) from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769605       1 reflector.go:257] Listing and watching *v1.PartialObjectMetadata from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769648       1 reflector.go:221] Starting reflector *v1.PartialObjectMetadata (9h55m40.357307208s) from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.769655       1 reflector.go:257] Listing and watching *v1.PartialObjectMetadata from pkg/mod/k8s.io/client-go@v0.26.3/tools/cache/reflector.go:169
I0428 09:57:01.775924       1 leaderelection.go:278] successfully renewed lease cert-manager/policy.cert-manager.io
I0428 09:57:01.871239       1 shared_informer.go:303] caches populated
I0428 09:57:01.871421       1 shared_informer.go:303] caches populated
I0428 09:57:01.871492       1 shared_informer.go:303] caches populated
I0428 09:57:01.871495       1 shared_informer.go:303] caches populated
I0428 09:57:01.871645       1 shared_informer.go:303] caches populated
I0428 09:57:01.871728       1 shared_informer.go:303] caches populated
I0428 09:57:01.871868       1 shared_informer.go:303] caches populated
I0428 09:57:01.871928       1 shared_informer.go:303] caches populated
I0428 09:57:01.871936       1 shared_informer.go:303] caches populated
I0428 09:57:01.871876       1 controller.go:228] controller-manager "msg"="Starting workers" "controller"="certificaterequestpolicy" "controllerGroup"="policy.cert-manager.io" "controllerKind"="CertificateRequestPolicy" "worker count"=1
I0428 09:57:01.871761       1 shared_informer.go:303] caches populated
I0428 09:57:01.872221       1 shared_informer.go:303] caches populated
I0428 09:57:01.871918       1 shared_informer.go:303] caches populated
I0428 09:57:01.871924       1 shared_informer.go:303] caches populated
I0428 09:57:01.872328       1 shared_informer.go:303] caches populated
I0428 09:57:01.872408       1 shared_informer.go:303] caches populated
I0428 09:57:01.872429       1 shared_informer.go:303] caches populated
I0428 09:57:01.872484       1 controller.go:228] controller-manager "msg"="Starting workers" "controller"="certificaterequest" "controllerGroup"="cert-manager.io" "controllerKind"="CertificateRequest" "worker count"=1
I0428 09:57:03.787115       1 leaderelection.go:278] successfully renewed lease cert-manager/policy.cert-manager.io
I0428 09:57:05.800596       1 leaderelection.go:278] successfully renewed lease cert-manager/policy.cert-manager.io
I0428 09:57:07.809900       1 leaderelection.go:278] successfully renewed lease cert-manager/policy.cert-manager.io
I0428 09:57:09.819435       1 leaderelection.go:278] successfully renewed lease cert-manager/policy.cert-manager.io
I0428 09:57:11.827727       1 leaderelection.go:278] successfully renewed lease cert-manager/policy.cert-manager.io
I0428 09:57:13.837715       1 leaderelection.go:278] successfully renewed lease cert-manager/policy.cert-manager.io
I0428 09:57:15.848863       1 leaderelection.go:278] successfully renewed lease cert-manager/policy.cert-manager.io
```
</detail>

File: Makefile (hunk 5 above)
Comment (jsoref) [reply Makefile:17]: This is how we've been using buildx:
https://github.com/GarnerCorp/build-actions/blob/075e5985b1deb54fd6cf733ad2bdbb9ac389274e/scripts/docker-build-push.sh#L44-L51

Comment (irbekrm) [reply Makefile]: Yeah I think in general we also should adapt that approach of setting up a builder as opposed to using whatever is the default configured builder on people's machines, I've had issues with this in the past.
This seems to be an overkill here, so I've removed the use of the buildx plugin.

**Comment by irbekrm** [comment]
Thanks both, @wallrj I've added the fixes suggested PTAL

> PS I noticed lots of duplicated log messages. See output below:

Those messages all come from aprover-policy, we don't start any controllers or leader election here.

**Review by wallrj — APPROVED** [review]
/lgtm

File: README.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# example-approver-policy-plugin
```
Comment (wallrj) [inline README.md:1]: @SgtCoDFish added the cert-manager badge to the top of all the other repos README files
Suggest we do the same here.

**Review by wallrj — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (irbekrm) [reply README.md:10]: Thanks Richard, that's a good idea, I've added it.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
+<p align="center">
+  <img src="https://raw.githubusercontent.com/cert-manager/cert-manager/d53c0b9270f8cd90d908460d69502694e1838f5f/logo/logo-small.png" height="256" width="256" alt="cert-manager project logo" />
+</p>
+<p align="center">
+  <a href="https://godoc.org/github.com/cert-manager/approver-policy"><img src="https://godoc.org/github.com/cert-manager/approver-policy?status.svg" alt="approver-policy godoc"></a>
+  <a href="https://goreportcard.com/report/github.com/cert-manager/approver-policy"><img alt="Go Report Card" src="https://goreportcard.com/badge/github.com/cert-manager/approver-policy" /></a>
+  <a href="https://artifacthub.io/packages/search?repo=cert-manager"><img alt="Artifact Hub" src="https://img.shields.io/endpoint?url=https://artifacthub.io/badge/repository/cert-manager" /></a>
+</p>
```
Comment (wallrj) [inline README.md:8]: These badges are for the different repos. Let's remove those.

```suggestion
```

**Review by wallrj — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (irbekrm) [reply README.md:8]: Ahh I need to actually place the logo file in this repo 💡 
Will do that in a sec

Comment (irbekrm) [reply README.md]: ignore the previous sentence
I've removed the badges should be ok now PTAL

**Review by wallrj — APPROVED** [review]
/lgtm

**Comment by irbekrm** [comment]
I thought this had merged, but apparently needs a Github approve- @wallrj could you please approve it?

**Review by wallrj — APPROVED** [review]
/lgtm

**Review by wallrj — APPROVED** [review]
Sorry....I didn't read your comment properly.

I thought cert-manager maintainers PRs were automatically marked as approved?

/approve
/lgtm

**Comment by SgtCoDFish** [comment]
> I thought cert-manager maintainers PRs were automatically marked as approved?

I think only if they're in OWNERS, which there isn't in this repo.  Might need to manually add an `approved` label to get this bootstrapped!

**Pull request merged**